### PR TITLE
V1.4-Update use case in DG

### LINKED ISSUE #249: Update use cases in DG
Add in missing use cases for new commands added.
Update use case to be more specific.
Standardise all use cases.

### PR DESCRIPTION
**Pull request by ZhiWei1010**
This commit updates the use cases in the DG
* Edit all use case to end at extensions instead of resume
* Add new use case for edit event, edit patient, find patient by keyword, sort, help, clear, exit.
* Standardise all use cases.

Fixes #249

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1066,54 +1066,116 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
-    * 2a1. PatientSync shows an error message.
+* 1a. Nurse's input is invalid command format.
 
+    * 1a1. PatientSync shows an error message.
+           
       Use case ends.
+  
+* 1b. Nurse's input contains invalid parameters value. 
 
+    * 1b1. PatientSync shows an error message.
```
Comment (choowengyan) [inline docs/DeveloperGuide.md:1101]: Perhaps can add one more use case where Nurse added patient with duplicated patient hospital ID?

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1066,54 +1066,116 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+    
+      Use case ends.
+
+**Use case: UC03 - Edit a patient**
+
+**MSS**
+
+1. Nurse requests to list patients.
+1. PatientSync shows a list of patients.
+1. Nurse requests to edit a specific patient in the list.
```
Comment (choowengyan) [inline docs/DeveloperGuide.md]: Would it be better to phrase it as `Nurse requests to edit a specific patient's information/ field in the list`?

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1066,54 +1066,116 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+1. PatientSync edit the patient and displays success message.
```
Comment (choowengyan) [inline docs/DeveloperGuide.md]: Would it be better to phrase it as` PatientSync edits the patient information and displays success message` ?

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1066,54 +1066,116 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+   Use case ends.
+
+**Extensions**
+
+* 2a. The list is empty.
   Use case ends.
 
-* 3a. The given index is invalid.
+* 3a. Nurse's input is invalid command format.
```
Comment (choowengyan) [inline docs/DeveloperGuide.md:1163]: Would be better if it is phrased to `Nurse's input command format is invalid`?

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1066,54 +1066,116 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
 
-    * 3a1. PatientSync shows an error message.
+  * 3a1. PatientSync shows an error message.
+
+    Use case ends.
+
+* 3b. Nurse's input contains invalid patient id.
```
Comment (choowengyan) [inline docs/DeveloperGuide.md]: Would suggest to change -> Nurse's input contains invalid ~~patient id~~ `patientHospitalId` / `patient's hospital ID` (depending on the other use cases format)

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1066,54 +1066,116 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+  * 3b1. PatientSync shows an error message.
+
+    Use case ends.
 
-**Use case: List all patients**
+* 3c. Nurse's input contains invalid parameters value.
+
+  * 3c1. PatientSync shows an error message.
+
+    Use case ends.
```
Comment (choowengyan) [inline docs/DeveloperGuide.md]: Would be great as well if we could add one more use case where the index to edit inputted by nurse is invalid? or do you think it is considered under the invalid command format use case?

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1163,20 +1247,26 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
 **MSS**
 
-1.  Nurse requests to list patients
-2.  PatientSync shows a list of patients
-3.  Nurse requests to add a tag to a specific patient in the list
-4.  PatientSync add a tag to the patient
+1. Nurse requests to list patients.
+1. PatientSync shows a list of patients.
+1. Nurse requests to add a tag to a specific patient in the list.
+1. PatientSync add a tag to the patient and displays success message.
```
Comment (choowengyan) [inline docs/DeveloperGuide.md]: Would be good if `add` change to `adds`

**Review by choowengyan — CHANGES_REQUESTED** [review]
Overall, great use cases documentation! Just some suggestions on the use cases for add/ edit patient.

File: docs/DeveloperGuide.md (hunk 5 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: the patient ID here should be referring to the id in the patient list, thus not `patientHospitalId`

Comment (choowengyan) [reply docs/DeveloperGuide.md]: Would it be better if we use index instead?

File: docs/DeveloperGuide.md (hunk 7 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: good catch!

File: docs/DeveloperGuide.md (hunk 6 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: I think it can be considered under invalid parameters!

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: good suggestion! changed

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: good suggestion! changed

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md:1101]: do you think it can be counted as invalid parameters value?

Comment (choowengyan) [reply docs/DeveloperGuide.md:1101]: ya I think should be can!

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1163,20 +1253,26 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+  * 3a1. PatientSync shows an error message.
+
+    Use case ends.
 
-      Use case resumes at step 2.
+* 3b. Nurse's input contains invalid patient index or event index.
+
+  * 3b1. PatientSync shows an error message.
+
+    Use case ends.
```
Comment (guohuang88) [inline docs/DeveloperGuide.md:1306]:
```suggestion
* 3c. Nurse's input contains invalid event name.
    * 3c1. PatientSync shows an error message.
    Use case resumes at step 2.
* 3d. Nurse's input contains invalid date or datetime format.
    * 3d1. PatientSync shows an error message.
    Use case resumes at step 2.
* 3e. Nurse's input contains duplicate event.
    * 3e1. PatientSync shows an error message.
    Use case resumes at step 2.
```
These are the other cases where there will be an error message.

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1066,54 +1066,122 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
 * 2a. The list is empty.
+      Use case ends.
+
+* 3a. Nurse's input is invalid command format.
+
+    * 3a1. PatientSync shows an error message.
+    
+      Use case ends.
+
+* 3b. Nurse's input contains invalid patient id.
```
Comment (choowengyan) [inline docs/DeveloperGuide.md]: should it be `ID`/ `Id` instead if `id`?

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1066,54 +1066,122 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
 
-* 3a. The given index is invalid.
+* 3a. Nurse's input is invalid command format.
 
-    * 3a1. PatientSync shows an error message.
+  * 3a1. PatientSync shows an error message.
+
+    Use case ends.
+
+* 3b. Nurse's input contains invalid patient id.
```
Comment (choowengyan) [inline docs/DeveloperGuide.md]: Similar to the comment above regarding `id`.

**Review by choowengyan — COMMENTED** [review]
Just some comments regarding id as they are different from the Id in line 1094. Perhaps we can consider standardising whether we want id or Id / ID

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1163,20 +1224,28 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
 **MSS**
 
-1.  Nurse requests to list patients
-2.  PatientSync shows a list of patients
-3.  Nurse requests to add a tag to a specific patient in the list
-4.  PatientSync add a tag to the patient
+1.  Nurse requests to list patients.
+1.  PatientSync shows a list of patients.
+1.  Nurse requests to edit an event for a specific patient in the list.
+1.  PatientSync edits an event the patient and displays success message.
```
Comment (guohuang88) [inline docs/DeveloperGuide.md]:
```suggestion
1.  PatientSync edits an event for the patient and displays success message.
```

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1232,12 +1315,123 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+  Use case ends.
+
+* 3a. Nurse's input is invalid command format.
 
     * 3a1. PatientSync shows an error message.
 
       Use case resumes at step 2.
 
+**Use case: UC11 - Find patients by keyword**
```
Comment (guohuang88) [inline docs/DeveloperGuide.md]:
```suggestion
**Use case: UC11 - Find patients by name**
```

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1232,12 +1315,123 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+**MSS**
+
+1. Nurse requests to list patients.
+1. PatientSync shows a list of patients.
+1. Nurse requests to find patients with specific keywords in the list.
+1. PatientSync finds patients with the specified keywords and displays success message.
```
Comment (guohuang88) [inline docs/DeveloperGuide.md]:
```suggestion
1. Nurse requests to find patients with specific name(s) in the list.
1. PatientSync finds patients with the specified name(s) and displays success message.
```

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1232,12 +1315,123 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+    Use case resumes at step 2.
+
+**Use case: UC10 - Find patients by a tag**
+
+**MSS**
+
+1. Nurse requests to list patients.
+1. PatientSync shows a list of patients.
+1. Nurse requests to find patients with a specific tag in the list.
+1. PatientSync finds patients with the specified tag and displays success message.
```
Comment (guohuang88) [inline docs/DeveloperGuide.md]:
```suggestion
1. Nurse requests to find patients with specific tag(s) in the list.
1. PatientSync finds patients with the specified tag(s) and displays success message.
```

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1232,12 +1315,123 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+
+    Use case resumes at step 2.
+
+* 3c. Nurse's input contains invalid tags.
+
+  * 3c1. PatientSync shows an error message.
+
+    Use case resumes at step 2.
+
+**Use case: UC10 - Find patients by a tag**
```
Comment (guohuang88) [inline docs/DeveloperGuide.md]:
```suggestion
**Use case: UC10 - Find patients by tag**
```

File: docs/DeveloperGuide.md (hunk 9 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: Good catch!

File: docs/DeveloperGuide.md (hunk 10 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: Good catch!

File: docs/DeveloperGuide.md (hunk 14 above)
Comment (guohuang88) [reply docs/DeveloperGuide.md]: User can find using 1 or more tags

**Review by guohuang88 — CHANGES_REQUESTED** [review]
Good job on rewriting the use cases! I have added some suggestions!

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1186,20 +1255,26 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
 
 **MSS**
 
-1.  Nurse requests to list patients
-2.  PatientSync shows a list of patients
-3.  Nurse requests to delete a tag from a specific patient in the list
-4.  PatientSync deletes a tag from the patient
+1. Nurse requests to list patients.
+1. PatientSync shows a list of patients.
+1. Nurse requests to add a tag to a specific patient in the list.
```
Comment (NeoHW) [inline docs/DeveloperGuide.md]: one or more tags

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1186,20 +1255,26 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+1. PatientSync adds a tag to the patient and displays success message.
```
Comment (NeoHW) [inline docs/DeveloperGuide.md]: adds the specified tag(s) and display status message

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1209,20 +1284,28 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
-**Use case: Find patients by a tag**
+* 3b. Nurse's input contains invalid patient index.
+
+  * 3b1. PatientSync shows an error message.
+
+    Use case resumes at step 2.
+
+**Use case: UC09 - Delete tag from a patient**
+
+**Preconditions: Tag exists for the target patient**
```
Comment (NeoHW) [inline docs/DeveloperGuide.md]: better not to use preconditions in use case and put it under extensions

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1232,12 +1315,123 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+  * 3b1. PatientSync shows an error message.
+
+    Use case resumes at step 2.
+
+* 3c. Nurse's input contains invalid tags.
+
+  * 3c1. PatientSync shows an error message.
+
+    Use case resumes at step 2.
+
```
Comment (NeoHW) [inline docs/DeveloperGuide.md:1400]: * 3d. The patient selected does not contain specified tag
* 3d1. PatientSync shows an error message.
Use case resumes at step 2

**Review by NeoHW — CHANGES_REQUESTED** [review]

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -1209,20 +1284,28 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
 
 **MSS**
 
-1.  Nurse requests to list patients
-2.  PatientSync shows a list of patients
-3.  Nurse requests to find patients with a specific tag in the list
-4.  PatientSync finds patients with the tag
+1. Nurse requests to list patients.
+1. PatientSync shows a list of patients.
+1. Nurse requests to delete a tag from a specific patient in the list.
```
Comment (NeoHW) [inline docs/DeveloperGuide.md]: one or more tags

File: docs/DeveloperGuide.md [hunk 21]
```diff
@@ -1209,20 +1284,28 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
+1. PatientSync deletes a tag from the patient and displays success message.
```
Comment (NeoHW) [inline docs/DeveloperGuide.md]: delete the specified tag(s) in the patient's tag list

**Review by NeoHW — CHANGES_REQUESTED** [review]

File: docs/DeveloperGuide.md (hunk 13 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: It is being consistent with the Section on locating patients by name 3.10. Maybe should standardise both?

File: docs/DeveloperGuide.md (hunk 12 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: It is being consistent with the Section on locating patients by name 3.10. Maybe should standardise both?

File: docs/DeveloperGuide.md (hunk 14 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: Nice catch! updated!

File: docs/DeveloperGuide.md (hunk 15 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: updated to tag(s)! to be consistent with below comments!

File: docs/DeveloperGuide.md (hunk 19 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md:1400]: Updated!

File: docs/DeveloperGuide.md (hunk 20 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: Updated!

File: docs/DeveloperGuide.md (hunk 21 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: Updated!

File: docs/DeveloperGuide.md (hunk 18 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: Good point! removed the precondition!

File: docs/DeveloperGuide.md (hunk 16 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: Updated !

File: docs/DeveloperGuide.md (hunk 17 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: Updated!

File: docs/DeveloperGuide.md (hunk 8 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md:1306]: Updated!

File: docs/DeveloperGuide.md (hunk 11 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: Good catch! updated!

File: docs/DeveloperGuide.md (hunk 12 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: After offline discussion, it is ok to change to name!

File: docs/DeveloperGuide.md (hunk 13 above)
Comment (ZhiWei1010) [reply docs/DeveloperGuide.md]: After offline discussion, it is ok to change to name!

**Review by choowengyan — APPROVED** [review]
LGTM!

**Review by guohuang88 — APPROVED** [review]

**Review by guohuang88 — COMMENTED** [review]
LGTM

**Review by NeoHW — APPROVED** [review]
lgtm

**Review by SherisseTJW — APPROVED** [review]
lgtm

**Pull request merged**